Various improvements to setup-homebrew

Bo98 opened the pull request:
Docker file ownership:
* Replace `chown` logic with `setfacl`.
  - This allows the original permissions to not be disrupted, which the runner may need (and does for the new file commands).
* Increase to cover `RUNNER_TEMP`, including file commands (e.g. `GITHUB_OUTPUT`).
* Reset directories to original ACL/permissions after job completes.

Tokens:
* Adjust token behaviour to use HTTP headers rather than URL authentication.
  - This matches how `actions/checkout` behaves.
* Cleanup token config after the job completes.
  - This is a security enchancement.
* Document that tokens persist through job steps.

GitHub workspace:
* Remove `GITHUB_WORKSPACE` replacement for Homebrew/brew and Homebrew/homebrew-core.
  - ⚠️ This is a breaking change affecting those repositories only.
* Adjust `GITHUB_WORKSPACE` handling for other taps to instead clone to the original `GITHUB_WORKSPACE` directory and symlink the tap location to it, rather than vice versa.
  - This has a slight compatibility concern, albeit hopefully not noticeable.
  - This fixes various compatibility issues with actions assuming `GITHUB_WORKSPACE` is untampered.
  - This operation is now cleaned up and undone when the job completes.
  - This behaviour continues (as before) to not apply to container jobs.
* Add `repository-path` output for Homebrew/brew and all taps, pointing to the relevant path matching `GITHUB_REPOSITORY`

Safe directory:
* Whitelist all Homebrew repositories in git's `safe.directory` config when running in a container.
  - This fixes errors when running not as the `linuxbrew` user.
  - This whitelisting action is currently only supported in container jobs.

Other:
* Replace `set-output` with `GITHUB_OUTPUT` file command.
* Move `sbin` to lower priority than `bin`.
  - This makes the behaviour consistent to `brew shellenv`.
* Remove Linux-specific permission changes (experimental - might drop this if `brew tests` fails).

Bo98 commented:
This ideally needs more coverage to actually test the token changes and various update paths involving different taps.

Bo98 commented:
> Remove `GITHUB_WORKSPACE` replacement for Homebrew/brew and Homebrew/homebrew-core.
> * ⚠️ This is a breaking change affecting those repositories only.

This will break most workflows in Homebrew/brew that assume pwd == repo. But it is equally necessary if we want to update from `actions/cache@v1`.

It's maybe worth adding a `repository-path` output here to make it easy to set `working-directory` in subsequent steps.

MikeMcQuaid commented on .github/workflows/setup-homebrew.yml:
```diff
@@ -11,15 +11,42 @@ on:
       - name: Checkout
-        uses: actions/checkout@v3
+        run: |
+          # Allow writing under linuxbrew user
+          orig_uid=$(ls -nd $GITHUB_WORKSPACE | awk 'NR==1 {print $3}')
+          sudo chown -R $(whoami) $GITHUB_WORKSPACE $HOME
+
+          git config --global --add safe.directory $GITHUB_WORKSPACE
+          git init
+          git remote add origin $GITHUB_SERVER_URL/$GITHUB_REPOSITORY
+          git fetch --no-tags --depth=1 origin $GITHUB_SHA
+          git checkout --force $GITHUB_SHA
+
+          # Reset ownership so we test if the equivalent in setup-homebrew works.
+          sudo chown -R $orig_uid $GITHUB_WORKSPACE $HOME
```
Can you explain a bit more what's going on here? I'm pretty lost.

Bo98 replied:
We never tested setup-homebrew under Homebrew Docker images in this repo's CI, so I thought it was about time we did.

`actions/checkout@v3` does not work with our Homebrew Docker images. The necessary permission fixes are included in our setup-homebrew, but the point of this CI is to checkout so we can test the setup-homebrew changes.

So this replaces `actions/checkout` with a simple checkout + the temporary permission change necessary for that to happen.

MikeMcQuaid replied:
> `actions/checkout@v3` does not work with our Homebrew Docker images. The necessary permission fixes are included in our setup-homebrew, but the point of this CI is to checkout so we can test the setup-homebrew changes.
> 
> So this replaces `actions/checkout` with a simple checkout + the temporary permission change necessary for that to happen.

Could this get added as an inline comment? Thanks ❤️

MikeMcQuaid commented on setup-homebrew/main.mjs:
```diff
@@ -0,0 +1,13 @@
+import {exec} from "@actions/exec"
+import core from "@actions/core"
+
+try {
+  await exec("bash", [
+    new URL("./main.sh", import.meta.url).pathname,
+    core.getInput("test-bot"),
+    core.getInput("debug"),
+    core.getInput("token")
+  ])
+} catch (error) {
+  core.setFailed(error.message)
+}
```
I don't really understand why this is needed, can you elaborate a bit?

Bo98 replied:
Unfortunantely, GitHub does not support `post` actions in shell so a JS wrapped is needed.

MikeMcQuaid commented on setup-homebrew/main.sh:
```diff
@@ -44,68 +44,96 @@ HOMEBREW_PREFIX="$(brew --prefix)"
 if [[ -f "/.dockerenv" ]] || ([[ -f /proc/1/cgroup ]] && grep -qE "actions_job|docker" /proc/1/cgroup); then
     # Fix permissions to give normal user access
-    sudo chown -R "$(whoami)" "$HOME" "$PWD/.."
+    if [[ -n "$(command -v setfacl)" ]]; then
+        setfacl_dirs=()
+        user=$(whoami)
+        for dir in "$HOME" "$RUNNER_WORKSPACE" "$RUNNER_TEMP"
+        do
+            if [[ ! -w "$dir" ]]; then
+                sudo setfacl -Rm "d:u:$user:rwX,u:$user:rwX" "$dir"
+                setfacl_dirs+=("$dir")
+            fi
+        done
+        echo "SETFACL_DIRECTORIES=$(IFS=:; echo "${setfacl_dirs[*]}")" >> $GITHUB_STATE
+        echo "Set up ACL."
+    elif [[ ! -w "$GITHUB_OUTPUT" ]]; then
+        echo "::warning::Missing write permissions to GitHub directories. Install \`acl\` (\`setfacl\`)."
+    fi
```
This is pretty involved. Can it get more comments explaining what's going on, why it's needed, etc?

Bo98 replied:
This is the exact same thinng as before with  `chown` but using `setfacl` instead, which works better as it can apply to new files (needed for file commads to work like `GITHUB_OUTPUT`).

`setfacl` _might_ not be installed so I added a check to be safe.

The only other thing happening here is to track what we've changed so that we can cleanup and revert back when the job completes.

MikeMcQuaid commented on setup-homebrew/main.sh:
```diff
@@ -44,68 +44,96 @@ HOMEBREW_PREFIX="$(brew --prefix)"
+        echo "::warning::Missing write permissions to GitHub directories. Install \`acl\` (\`setfacl\`)."
+    fi
+
+    # Add safe directories.
```
What are safe directories? Again: can this get more comments explaining.

Bo98 replied:
Because of https://github.blog/2022-04-12-git-security-vulnerability-announced/, running as root and performing git actions for a directory owned by `linuxbrew` results in an error without this.

See https://github.com/actions/checkout/commit/dcd71f646680f2efd8db4afa5ad64fdcba30e748 for the equivalent change GitHub did for `actions/checkout`.

MikeMcQuaid replied:
Sorry, still don't understand what behaviour this changes.

MikeMcQuaid replied:
Thanks! When is this called as root?

Bo98 replied:
For our own Docker images: we mostly avoid it ourselves except https://github.com/Homebrew/actions/blob/master/.github/workflows/setup-commit-signing.yml#L20.

For non-Homebrew Docker images: most of them run as root. The `--user=linuxbrew` we have is unusual for the most part (but makes sense for Homebrew/core etc.)

MikeMcQuaid replied:
Ok, thanks.

> GitHub only actually support root in Docker: https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions#user, so it makes sense to make this work as root.

Note at least internal to GitHub: the opposite is the case 😄.

Given no-one has complained about this I'm fine with including it but don't feel like it's super high priority?

Bo98 replied:
> Note at least internal to GitHub: the opposite is the case 😄.

`actions/checkout@v3.1.0` no longer works as non-root anymore. It can't be used with our Homebrew Docker images anymore, hence the `setup-homebrew.yml` CI change.

MikeMcQuaid commented on setup-homebrew/main.sh:
```diff
@@ -138,14 +163,15 @@ else
-                ln -vs "$HOMEBREW_TAP_REPOSITORY" "$GITHUB_WORKSPACE"
-            fi
+
+        # This is mostly kept for compatibility reasons, so the conditions are intentionally restrictive.
```
Can this comment explain more what's going on here

MikeMcQuaid commented on setup-homebrew/post.mjs:
```diff
@@ -0,0 +1,11 @@
+import {exec} from "@actions/exec"
+import core from "@actions/core"
+
+try {
+  await exec("bash", [
+    new URL("./post.sh", import.meta.url).pathname,
+    core.getInput("debug")
+  ])
+} catch (error) {
+  core.setFailed(error.message)
+}
```
What's this needed for?

Bo98 replied:
Unfortunantely, GitHub does not support `post` actions in shell so a JS wrapped is needed.

MikeMcQuaid replied:
Why do we need post actions now?

Bo98 replied:
See post.sh review below.

MikeMcQuaid reviewed (commented):
Fine with this once it has a few more comments.

Would be great to explain the motivation for the changes rather than just what's changing. I don't really understand the changes here.

Is `$PWD` still the "correct" repository here but it may be a symlink instead? If not: what is `$PWD` now?

Would it be worth splitting this into multiple PRs? This feels like a lot of things to change at once given how many other repositories rely on this.

Bo98 commented:
> Is `$PWD` still the "correct" repository here but it may be a symlink instead? If not: what is `$PWD` now?

`PWD` by default it always `GITHUB_WORKSPACE`. `GITHUB_WORKSPACE` is a fixed location which cannot be changed (but the default working directory can be in the workflow files).

Homebrew and its taps are at fixed locations outside of `GITHUB_WORKSPACE`. To work around the two things being  immovable, we replaced `GITHUB_WORKSPACE` with a symlink to the Homebrew-installed location. This worked for fixing `PWD` but actively broke some official GitHub actions such as `actions/cache@v2` (and later).

We cannot stay on `actions/cache@v1` forever and fight against what GitHub support, and they do not support replacing `GITHUB_WORKSPACE`. The symlink also breaks runner code on non-ephemeral (hence the self-hosted checks that were here).

The alternative is to reverse this so that we always checkout to `GITHUB_WORKSPACE` but symlink on the Homebrew side. As long as Homebrew itself supports symlinks, this should fix all the issues.

Symlinking Homebrew/brew in that direction is however tricky given the nested GitHub repositories. I decided to just drop support for replacing `GITHUB_WORKSPACE` for that repo, given that breakage only affects us and can be addressed in our workflow files. So for Homebrew/brew, `PWD` would no longer point to the repo.

For Homebrew/core, we could keep it. I dropped it considering it was never enabled for self-hosted and our workflows in Homebrew/core need to work on self-hosted, and there is only one thing I think needs changing. I think consistency  between runners makes sense.

For other taps though, I have got the reversed symlink in place with this PR. I have kept the original restrictions of not applying to containers etc. I don't think the restrictions are necessary anymore, but relaxing them does mean making a promise to never reverse that decision, and I want to avoid those promises for now.

MikeMcQuaid commented:
> The alternative is to reverse this so that we always checkout to `GITHUB_WORKSPACE` but symlink on the Homebrew side. As long as Homebrew itself supports symlinks, this should fix all the issues.

👍🏻 makes sense.

> So for Homebrew/brew, `PWD` would no longer point to the repo.

I think this is going to cause a bit of confusion, longer-term, and ideally we'd address it. Perhaps a Homebrew/brew environment variable to avoid the symlink-resolving prefix logic would help here?

> For Homebrew/core, we could keep it.

This would be good 👍🏻 

> I think consistency between runners makes sense.

Agreed 👍🏻

Bo98 commented on setup-homebrew/post.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+set -euo pipefail
+
+DEBUG="${1}"
+
+if [[ "${DEBUG}" == 'true' ]]; then
+  set -x
+fi
+
+if [[ -n "${STATE_TAP_SYMLINK-}" ]]; then
+  rm "${STATE_TAP_SYMLINK}"
+  mkdir "${STATE_TAP_SYMLINK}"
+  (shopt -s dotglob; mv "${GITHUB_WORKSPACE}/*" "${STATE_TAP_SYMLINK}" )
+  echo "Reset tap symlink."
+fi
+
+if [[ -n "${STATE_TOKEN_SET-}" ]]; then
+  git config --global --unset-all "http.${GITHUB_SERVER_URL}/.extraheader"
+  echo "Removed token."
+fi
+
+if [[ -n "${STATE_SETFACL_DIRECTORIES-}" ]]; then
+  (IFS=:; sudo setfacl -Rb ${STATE_SETFACL_DIRECTORIES})
+  echo "Reset permissions."
+fi
```
This is a cleanup step to restore everything back to how it was before the job started.

* Tokens were previously never cleaned up. This meant subsequent jobs could read them. I consider this to be a security vulnerability.
* Symlinks should be reverted otherwise if the same runnner is used for two taps then we might have symlinks for two taps pointing to the same place.
* `setfacl` reverting I think is good to do. Probably won't _break_ if not done, but I still think it's good to clean up.

MikeMcQuaid replied:
Ok, I see, so this is for cleanup after all of the job has completed and not just `setup-homebrew`.

Good to me, then. Again: would like a lot more inline comments explaining what's going on to future readers.

Bo98 commented:
> I think this is going to cause a bit of confusion, longer-term, and ideally we'd address it. Perhaps a Homebrew/brew environment variable to avoid the symlink-resolving prefix logic would help here?

What type of confusion?

The assumption that PWD == repo already doesn't apply to most of Homebrew/core because that's largely self-hosted.

Note that the code removed was gated under a `Homebrew/brew` check anyway, though that's because we had a bit of code duplication.

MikeMcQuaid commented:
> > I think this is going to cause a bit of confusion, longer-term, and ideally we'd address it. Perhaps a Homebrew/brew environment variable to avoid the symlink-resolving prefix logic would help here?
> 
> What type of confusion?

Confusion that Homebrew/brew is different to every other repository in this respect and different to what `actions/checkout` does for you.

> The assumption that PWD == repo already doesn't apply to most of Homebrew/core because that's largely self-hosted.

Why is it different there? Can it not be?

Bo98 commented:
> Why is it different there? Can it not be?

Because the `GITHUB_WORKSPACE` replacement broke things. Beyond `actions/checkout`, it broke the GitHub Action's runner's own code on self-hosted.

At this point making PWD == repo would require changes in various places. We currently write to the PWD in test-bot etc and this would break `brew doctor` as a result.

I'm a little hesistent to also do this on non-ephemeral in case the machine crashes mid-job.

Bo98 commented:
Added some comments.

MikeMcQuaid commented on setup-homebrew/action.yml:
```diff
@@ -14,20 +14,17 @@ inputs:
     default: false
   token:
-    description: Token to be used for GitHub authentication (if using private repositories)
+    description: Token to be used for GitHub authentication (if using private repositories). This token will persist through other steps for the duration of the job.
```
```suggestion
    description: Token to be used for GitHub authentication (if using private repositories). This token will only persist through other steps for the duration of the job.
```

Maybe?

Bo98 replied:
Possibly, but the tone of this is intended to be more of a security warning, to let the user know that the token could be read by subsequent steps (in case there's any untrusted code being run), rather than it being scoped to just setup-homebrew.

Bo98 replied:
1) Yes, with an opt-out (`persist-credentials`). Though they also apply the token by default unlike us.
2) action.yml states `The post-job step removes the PAT.`. The better description is at the near the start of the README: https://github.com/actions/checkout/blob/main/README.md

MikeMcQuaid replied:
Fine with me either way. Let's emulate what they do as much as possible, I reckon, because we're sort of a replacement for that for taps.

MikeMcQuaid commented on setup-homebrew/main.sh:
```diff
@@ -44,68 +44,101 @@ HOMEBREW_PREFIX="$(brew --prefix)"
 if [[ -f "/.dockerenv" ]] || ([[ -f /proc/1/cgroup ]] && grep -qE "actions_job|docker" /proc/1/cgroup); then
-    # Fix permissions to give normal user access
-    sudo chown -R "$(whoami)" "$HOME" "$PWD/.."
+    # Fix permissions to give container user access.
```
```suggestion
    # Fix permissions to give container user write access.
```

If correct?

MikeMcQuaid commented on setup-homebrew/main.sh:
```diff
@@ -44,68 +44,101 @@ HOMEBREW_PREFIX="$(brew --prefix)"
+        echo "Set up ACL."
+    elif [[ ! -w "$GITHUB_OUTPUT" ]]; then
+        # setfacl isn't installed on Ubuntu by default.
+        echo "::warning::Missing write permissions to GitHub directories. Install \`acl\` (\`setfacl\`)."
```
```suggestion
        echo '::warning::Missing write permissions to GitHub directories. Install `acl` (`setfacl`).'
```

MikeMcQuaid commented on setup-homebrew/main.sh:
```diff
@@ -138,14 +168,16 @@ else
-            fi
+
+        # Make repo available under `GITHUB_WORKSPACE` (default working directory), which some third-party taps may need.
+        # The symlink needs to be in this direction or `actions/cache` etc will break.
```
```suggestion
        # The symlink needs to be in this direction or `actions/cache` etc. will break as they rely on `GITHUB_WORKSPACE` being `PWD`
```

or whatever is correct

Bo98 commented:
Will merge this when I'm sure I'll be around to monitor CI afterwards. Today's the last day of the holidays (in Scotland) so probably tomorrow.
